compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,155 @@
+use super::MirPass;
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_index::IndexSlice;
+use rustc_middle::mir::traversal;
+use rustc_middle::mir::{
+    AggregateKind, BasicBlock, Body, Operand, Rvalue, StatementKind, TerminatorKind,
+};
+use rustc_middle::ty::TyCtxt;
+use rustc_session::Session;
+
+use super::simplify::remove_dead_blocks;
+
+/// This pass looks to optimize a pattern in MIR where a discriminant is
+/// constructed in one or more blocks with the same successor and then
+/// that discriminant is switched on in that successor block, in which case
+/// we can remove the need to switch on the discriminant at all because we
+/// statically know what target block will be taken for each constructed
+/// discriminant.
+///
+/// For example, the following CFG:
+/// ```text
+///     x = Foo::A(y); ---       Foo::A ---> ...
+///    /                  \            /
+/// ...                    --> switch x
+///    \                  /            \
+///     x = Foo::B(y); ---       Foo::B ---> ...
+/// ```
+/// would become:
+/// ```text
+///     x = Foo::A(y); --------- Foo::A ---> ...
+///    /
+/// ...
+///    \
+///     x = Foo::B(y); --------- Foo::B ---> ...
+/// ```
+pub struct SimplifyStaticSwitch;
+
+impl<'tcx> MirPass<'tcx> for SimplifyStaticSwitch {
+    fn name(&self) -> &'static str {
+        "SimplifyStaticSwitch"
+    }
+
+    fn is_enabled(&self, sess: &Session) -> bool {
+        sess.mir_opt_level() >= 2
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        debug!(body = %tcx.def_path_str(body.source.def_id()));
```

Comment:
```suggestion
        debug!(body = ?body.source.def_id());
```

`def_path_str` dooms the current compilation session.